src/graph/immutable_graph.cc

```diff
@@ -144,328 +97,112 @@ CSR::CSR(IdArray indptr, IdArray indices, IdArray edge_ids, bool is_multigraph,
-  IdArray eid = NewIdArray(len);
-  dgl_id_t* src_data = static_cast<dgl_id_t*>(src->data);
-  dgl_id_t* dst_data = static_cast<dgl_id_t*>(dst->data);
-  dgl_id_t* eid_data = static_cast<dgl_id_t*>(eid->data);
-  std::fill(src_data, src_data + len, vid);
-  std::copy(indices_data + off, indices_data + off + len, dst_data);
-  std::copy(edge_ids_data + off, edge_ids_data + off + len, eid_data);
-  return CSR::EdgeArray{src, dst, eid};
+  IdArray ret_dst = aten::CSRGetRowColumnIndices(adj_, vid);
+  IdArray ret_eid = aten::CSRGetRowData(adj_, vid);
+  IdArray ret_src = aten::Full(vid, ret_dst->shape[0], NumBits(), ret_dst->ctx);
+  return CSR::EdgeArray{ret_src, ret_dst, ret_eid};
 }
 
 CSR::EdgeArray CSR::OutEdges(IdArray vids) const {
   CHECK(IsValidIdArray(vids)) << "Invalid vertex id array.";
-  const dgl_id_t* indptr_data = static_cast<dgl_id_t*>(indptr_->data);
-  const dgl_id_t* indices_data = static_cast<dgl_id_t*>(indices_->data);
-  const dgl_id_t* edge_ids_data = static_cast<dgl_id_t*>(edge_ids_->data);
-  const auto len = vids->shape[0];
-  const dgl_id_t* vid_data = static_cast<dgl_id_t*>(vids->data);
-  int64_t rstlen = 0;
-  for (int64_t i = 0; i < len; ++i) {
-    dgl_id_t vid = vid_data[i];
-    CHECK(HasVertex(vid)) << "Invalid vertex: " << vid;
-    rstlen += OutDegree(vid);
-  }
-  IdArray src = NewIdArray(rstlen);
-  IdArray dst = NewIdArray(rstlen);
-  IdArray eid = NewIdArray(rstlen);
-  dgl_id_t* src_ptr = static_cast<dgl_id_t*>(src->data);
-  dgl_id_t* dst_ptr = static_cast<dgl_id_t*>(dst->data);
-  dgl_id_t* eid_ptr = static_cast<dgl_id_t*>(eid->data);
-  for (int64_t i = 0; i < len; ++i) {
-    const dgl_id_t vid = vid_data[i];
-    const dgl_id_t off = indptr_data[vid];
-    const int64_t deg = OutDegree(vid);
-    if (deg == 0)
-      continue;
-    const auto *succ = indices_data + off;
-    const auto *eids = edge_ids_data + off;
-    for (int64_t j = 0; j < deg; ++j) {
-      *(src_ptr++) = vid;
-      *(dst_ptr++) = succ[j];
-      *(eid_ptr++) = eids[j];
-    }
-  }
-  return CSR::EdgeArray{src, dst, eid};
+  auto csrsubmat = aten::CSRSliceRows(adj_, vids);
+  auto coosubmat = aten::CSRToCOO(csrsubmat, false);
+  // Note that the row id in the csr submat is relabled, so
+  // we need to recover it using an index select.
+  auto row = aten::IndexSelect(vids, coosubmat.row);
+  return CSR::EdgeArray{row, coosubmat.col, coosubmat.data};
 }
 
 DegreeArray CSR::OutDegrees(IdArray vids) const {
   CHECK(IsValidIdArray(vids)) << "Invalid vertex id array.";
-  const auto len = vids->shape[0];
-  const dgl_id_t* vid_data = static_cast<dgl_id_t*>(vids->data);
-  DegreeArray rst = DegreeArray::Empty({len}, vids->dtype, vids->ctx);
-  dgl_id_t* rst_data = static_cast<dgl_id_t*>(rst->data);
-  for (int64_t i = 0; i < len; ++i) {
-    const auto vid = vid_data[i];
-    CHECK(HasVertex(vid)) << "Invalid vertex: " << vid;
-    rst_data[i] = OutDegree(vid);
-  }
-  return rst;
+  return aten::CSRGetRowNNZ(adj_, vids);
```

Comment:
Are we missing the implementation for `uint64_t OutDegree(dgl_id_t vid)`?

Reply:
Here: https://github.com/dmlc/dgl/blob/a1a355f7e7c6437abff39d8be1aa011ab4c4723f/include/dgl/immutable_graph.h#L149